stonesoup/initiator/composite.py

```diff
@@ -0,0 +1,93 @@
+# -*- coding: utf-8 -*-
+from typing import Sequence
+
+from .base import Initiator
+from ..base import Property
+from ..types.detection import MissedDetection
+from ..types.hypothesis import SingleHypothesis, CompositeHypothesis
+from ..types.state import CompositeState
+from ..types.track import Track
+from ..types.update import CompositeUpdate
+
+
+class CompositeUpdateInitiator(Initiator):
+    """A composition of initiators, initiating tracks in a composite state space.
+
+    Utilises its sub-initiators to attempt to initiate a track for a :class:`~.CompositeDetection`
+    in each corresponding sub-state space, then combines the resultant track states in to a
+    :class:`~.CompositeState` track. If any sub-state is missing from the detection, or a
+    sub-initiator fails to initiate in its sub-state space, the corresponding sub-state of the
+    :attr:`prior_state` will be used instead.
+    """
+    sub_initiators: Sequence[Initiator] = Property()
+    prior_state: CompositeState = Property(
+        default=None,
+        doc="Prior state information. Defaults to `None`, in which case all sub-initiators must "
+            "have their own `prior_state` attribute.")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.prior_state is None:
+            for sub_initiator in self.sub_initiators:
+                try:
+                    sub_prior = getattr(sub_initiator, 'prior_state')
+                except AttributeError:
+                    raise ValueError("If no default prior state is defined, all sub-initiators "
+                                     "must have a `prior_state` attribute")
+                else:
+                    if sub_prior is None:
+                        raise ValueError("If no default prior state is defined, all "
+                                         "sub-initiators require a defined `prior_state`")
+
+    @property
+    def _prior_state(self):
+        if self.prior_state:
+            return self.prior_state
+        else:
+            return [sub_initiator.prior_state for sub_initiator in self.sub_initiators]
+
+    def initiate(self, detections, timestamp, **kwargs):
+        tracks = set()
+
+        for detection in detections:
+
+            mapping = detection.mapping
+
+            hypotheses = list()
+            states = list()
+
+            for i, sub_initiator in enumerate(self.sub_initiators):
+
+                try:
+                    # Check if detection has sub-detection for this state index
+                    detection_index = mapping.index(i)
+
+                    # Get sub-detection and initiate a (sub)track with it
+                    sub_detection = detection[detection_index]
+                    sub_tracks = sub_initiator.initiate({sub_detection}, timestamp=timestamp)
```

Comment:
I'm wondering if you should group all relevant detections together, and therefore call each sub initiator once. Initiation approach may be somehow reliant on the assumption that all detections from one sensor at the same timestamp are present.

Reply:
Agreed. That would look similar then to the composite hypothesiser method.